lib/Optimizer/Transforms/DependencyAnalysis.cpp

```diff
@@ -0,0 +1,3148 @@
+    return false;
+  }
+
+  // TODO: function arguments aren't really supported properly
+  //       in places like `OpDependencyNode::erase` or when handling
+  //       shadow dependencies, especially classical arguments.
+  //       I think function arguments shouldn't be supported and a
+  //       check should be made here, though the above issues could
+  //       be addressed and then they may be supported ok.
+
+  return true;
+}
+
+class DependencyAnalysisEngine {
+private:
+  SmallVector<DependencyNode *> perOp;
+  DenseMap<BlockArgument, ArgDependencyNode *> argMap;
+  SmallVector<Operation *> ifStack;
+  DenseMap<Operation *, SetVector<ShadowDependencyNode *>> freeClassicals;
+  unsigned vallocs;
+
+public:
+  DependencyAnalysisEngine()
+      : perOp({}), argMap({}), ifStack({}), freeClassicals({}), vallocs(0) {}
+
+  /// Creates a new dependency block for \p b by constructing a dependency graph
+  /// for the body of \p b starting from the block terminator.
+  ///
+  /// Any operation not somehow connected to the block terminator (this will
+  /// only happen if the entirety of a wire is irrelevant to the block
+  /// terminator, in which case it is considered to be dead code)
+  DependencyBlock *
+  visitBlock(mlir::Block *b,
+             SmallVector<DependencyNode::DependencyEdge> dependencies) {
+    SmallVector<ArgDependencyNode *> argdnodes;
+    for (auto targ : b->getArguments()) {
+      ArgDependencyNode *dnode;
+      // Entry block has no argument dependencies
+      if (dependencies.size() > 0)
+        dnode =
+            new ArgDependencyNode(targ, dependencies[targ.getArgNumber() + 1]);
+      else
+        dnode = new ArgDependencyNode(targ);
+      argMap[targ] = dnode;
+      argdnodes.push_back(dnode);
+    }
+
+    DenseMap<DependencyNode *, Operation *> roots;
+    TerminatorDependencyNode *terminator = nullptr;
+    for (auto &op : b->getOperations()) {
+      bool isTerminator = (&op == b->getTerminator());
+      auto node = visitOp(&op, isTerminator);
+
+      if (!node)
+        return nullptr;
+
+      if (isa<quake::ReturnWireOp>(&op))
+        roots[node] = &op;
+
+      if (isTerminator) {
+        assert(op.hasTrait<mlir::OpTrait::IsTerminator>() &&
+               "Illegal terminator op!");
+        terminator = static_cast<TerminatorDependencyNode *>(node);
+      }
+    }
+
+    DependencyGraph *new_graph = new DependencyGraph(terminator);
+    auto included = new_graph->getRoots();
+
+    // In debug mode, alert about dead code wires
+    // TODO: If an unused wire flows through an `if` with a useful wire,
+    //       then the unused wire is considered useful as the parent context
+    //       doesn't know that it doesn't interact with anything inside the if,
+    //       it would be nice to have a "hasInteraction" predicate inside `if`s
+    //       to be able to detect this case and do a better job of removing
+    //       unused wires.
```

Comment:
Do you actually see such a structure being created?

memtoreg should be generating a pruned SSI form for any wire values. That means any wire value that is live across the `if` but not used within the `if` wouldn't be part of the $\sigma$ or $\phi$ nodes of the `if` by construction.

That said, it's possible that ops get moved around and break the pruned SSI property.

Reply:
The situation I was (trying to) describe is a little trickier. Consider the following program
```
cudaq::qubit q,p;

if (some_condition) {
  x(p);
  x(q);
} else {
  y(p);
  y(q);
}

return mz(q);
```
Here, both `q` and `p` flow into (and are used inside) the `if`, so `p` is not treated as dead code, but it is actually dead code (well, according to DependencyAnalysis' definition of dead code) because it has no impact on the return value. This should be detectable (at least in a simpler case like this).

Reply:
If that helps to clarify the TODO, I can add the example to the TODO?

Reply:
OK.

We have to stay at the same level of abstraction to discuss the point here.
```mlir
  p = alloca
  q = alloca
  [p', q'] = cc.if (...) (p, q) {
    // p,q used
  } else {
    // p,q used
 }
 mz q'
 dealloc p'  // this is not optional and a use
 dealloc q'
 return
}
```
Because of the dealloc, the only correct representation necessitates that both p and q remain alive across the entire `if`.

Now, the dealloc could be replicated in both arms of the if which would sink that operation into the `if`, but doesn't seem to optimize anything to me.

Reply:
The issue is that `if` uses both `p` and `q`, producing `p'` and `q'`, so the dependency graph will start searching from the return, see the operation on `p'`, and then see the `if` producing `p'`, from which it will see `q'`, so it will include `q'` in the graph, as if the `if` was any other quake operation which entangled `p` and `q`, but because it's an `if`, `p` and `q` may or may not interact in the body of the `if`, so it's not always the case that the graph needs to chase down all wires flowing through an `if` (there's no harm in doing so, it is just suboptimal).

Reply:
OK.

Yes, the `if` must consider both arms in aggregate, and there has to be both arms. But, by virtue of this being a _conditional_ operation by definition, it need not be the case all qubits are used unconditionally. heh

Coincidentally, we can consider qubits in control positions to be "unused" in the sense that their quantum state is unchanged.

Reply:
> If that helps to clarify the TODO, I can add the example to the TODO?

We can turn this thread into an issue.

Reply:
Will create a separate issue for this one (or add to a new grouped issue).

Reply:
https://github.com/NVIDIA/cuda-quantum/issues/2181 written. Resolving this conversation.